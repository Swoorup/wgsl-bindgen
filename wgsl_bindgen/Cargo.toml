[package]
name = "wgsl_bindgen"
authors.workspace = true
version.workspace = true
edition.workspace = true
repository.workspace = true
documentation.workspace = true
license.workspace = true
description = "Type safe Rust bindings workflow for wgsl shaders in wgpu"
readme = "../README.md"

[dependencies]
naga = { workspace = true, features = ["wgsl-in"] }
wgpu-types.workspace = true
syn.workspace = true
quote.workspace = true
proc-macro2.workspace = true
prettyplease.workspace = true
thiserror.workspace = true
case.workspace = true
naga_oil.workspace = true
regex.workspace = true
data-encoding.workspace = true
indexmap.workspace = true
smallvec.workspace = true
derive_more.workspace = true
fxhash.workspace = true
derivative.workspace = true
smol_str.workspace = true
colored.workspace = true
derive_builder.workspace = true
miette.workspace = true
blake3.workspace = true
regex-syntax.workspace = true
strum.workspace = true
strum_macros.workspace = true
pathdiff.workspace = true
enumflags2.workspace = true
heck.workspace = true

[dev-dependencies]
indoc.workspace = true
pretty_assertions.workspace = true
miette = { workspace = true, features = ["fancy", "syntect"] }
