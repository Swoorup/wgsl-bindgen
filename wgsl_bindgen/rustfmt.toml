indent_style = "Block"
reorder_imports = true
tab_spaces = 2
max_width=90
fn_call_width=80
group_imports="StdExternalCrate"
