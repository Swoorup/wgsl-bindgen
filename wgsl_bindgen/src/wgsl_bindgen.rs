use std::io::Write;
use std::path::PathBuf;

use derive_builder::Builder;
use miette::Diagnostic;
use naga_oil::compose::{
  ComposableModuleDescriptor, Composer, ComposerError, NagaModuleDescriptor,
  ShaderLanguage,
};
use thiserror::Error;

use self::source_file::SourceFile;
use crate::{
  bevy_util::*, WgslEntryResult, WgslTypeMap, WgslTypeMapBuild, WgslTypeSerializeStrategy,
};
use crate::{create_rust_bindings, CreateModuleError, SourceFilePath};

const PKG_VER: &str = env!("CARGO_PKG_VERSION");
const PKG_NAME: &str = env!("CARGO_PKG_NAME");

/// Enum representing the possible errors that can occur in the `wgsl_bindgen` process.
///
/// This enum is used to represent all the different kinds of errors that can occur
/// when parsing WGSL shaders, generating Rust bindings, or performing other operations
/// in `wgsl_bindgen`.
#[derive(Debug, Error, Diagnostic)]
pub enum WgslBindgenError {
  #[error("All required fields need to be set upfront: {0}")]
  OptionBuilderError(#[from] WgslBindgenOptionBuilderError),

  #[error(transparent)]
  #[diagnostic(transparent)]
  DependencyTreeError(#[from] DependencyTreeError),

  #[error("Failed to compose modules with entry `{entry}`\n{msg}")]
  NagaModuleComposeError {
    entry: String,
    msg: String,
    inner: naga_oil::compose::ComposerErrorInner,
  },

  #[error(transparent)]
  ModuleCreationError(#[from] CreateModuleError),

  #[error(transparent)]
  WriteOutputError(#[from] std::io::Error),

  #[error("Output file is not specified. Maybe use `generate_string` instead")]
  OutputFileNotSpecified,
}

#[derive(Default, Debug, Clone, Copy, PartialEq, Eq)]
pub enum WgslShaderSourceOutputType {
  /// Include the final shader string directly in the output
  #[default]
  FinalShaderString,

  /// Use Composer including helper functions which will be executed on runtime
  Composer,
}

/// A struct representing a directory to scan for additional source files.
///
/// This struct is used to represent a directory to scan for additional source files
/// when generating Rust bindings for WGSL shaders. The `module_import_root` field
/// is used to specify the root prefix or namespace that should be applied to all
/// shaders given as the entrypoints, and the `directory` field is used to specify
/// the directory to scan for additional source files.
#[derive(Debug, Clone, Default)]
pub struct AdditionalScanDirectory {
  pub module_import_root: Option<String>,
  pub directory: String,
}

impl From<(Option<&str>, &str)> for AdditionalScanDirectory {
  fn from((module_import_root, directory): (Option<&str>, &str)) -> Self {
    Self {
      module_import_root: module_import_root.map(ToString::to_string),
      directory: directory.to_string(),
    }
  }
}

#[derive(Debug, Default, Builder)]
#[builder(
  setter(into),
  field(private),
  build_fn(private, name = "fallible_build")
)]
pub struct WgslBindgenOption {
  /// A vector of entry points to be added. Each entry point is represented as a `String`.
  #[builder(setter(each(name = "add_entry_point", into)))]
  pub entry_points: Vec<String>,

  /// The root prefix/namespace if any applied to all shaders given as the entrypoints.
  #[builder(default, setter(strip_option, into))]
  pub module_import_root: Option<String>,

  /// A boolean flag indicating whether to emit a rerun-if-changed directive to Cargo. Defaults to `true`.
  #[builder(default = "true")]
  pub emit_rerun_if_change: bool,

  /// A boolean flag indicating whether to skip header comments. Enabling headers allows to not rerun if contents did not change.
  #[builder(default = "false")]
  pub skip_header_comments: bool,

  /// A boolean flag indicating whether to skip the hash check. This will avoid reruns of bindings generation if
  /// entry shaders including their imports has not changed. Defaults to `false`.
  #[builder(default = "false")]
  pub skip_hash_check: bool,

  /// Derive [encase::ShaderType](https://docs.rs/encase/latest/encase/trait.ShaderType.html#)
  /// for user defined WGSL structs when `WgslTypeSerializeStrategy::Encase`.
  /// else derive bytemuck
  #[builder(default)]
  pub serialization_strategy: WgslTypeSerializeStrategy,

  /// Derive [serde::Serialize](https://docs.rs/serde/1.0.159/serde/trait.Serialize.html)
  /// and [serde::Deserialize](https://docs.rs/serde/1.0.159/serde/trait.Deserialize.html)
  /// for user defined WGSL structs when `true`.
  #[builder(default = "false")]
  pub derive_serde: bool,

  /// The type of output for the shader source. Defaults to `FinalShaderString`.
  #[builder(default)]
  pub shader_source_output_type: WgslShaderSourceOutputType,

  /// A mapping operation for WGSL built-in types. This is used to map WGSL built-in types to their corresponding representations.
  #[builder(setter(custom))]
  pub wgsl_type_map: WgslTypeMap,

  /// The output file path for the generated Rust bindings. Defaults to `None`.
  #[builder(default, setter(strip_option, into))]
  pub output_file: Option<PathBuf>,

  /// The additional set of directories to scan for source files.
  #[builder(default, setter(into, each(name = "additional_scan_dir", into)))]
  pub additional_scan_dirs: Vec<AdditionalScanDirectory>,
}

impl WgslBindgenOptionBuilder {
  pub fn build(&self) -> Result<WGSLBindgen, WgslBindgenError> {
    let options = self.fallible_build()?;
    WGSLBindgen::new(options)
  }

  pub fn wgsl_type_map(&mut self, map_build: impl WgslTypeMapBuild) -> &mut Self {
    let serialization_strategy = self
      .serialization_strategy
      .expect("Serialization strategy must be set before `wgs_type_map`");

    self.wgsl_type_map = Some(map_build.build(serialization_strategy));
    self
  }
}

pub struct WGSLBindgen {
  dependency_tree: DependencyTree,
  options: WgslBindgenOption,
  content_hash: String,
}

impl WGSLBindgen {
  fn new(options: WgslBindgenOption) -> Result<Self, WgslBindgenError> {
    let entry_points = options
      .entry_points
      .iter()
      .cloned()
      .map(SourceFilePath::new)
      .collect();

    let dependency_tree = DependencyTree::try_build(
      options.module_import_root.clone(),
      entry_points,
      options.additional_scan_dirs.clone(),
    )?;

    let content_hash = Self::get_contents_hash(&options, &dependency_tree);

    if options.emit_rerun_if_change {
      for file in Self::iter_files_to_watch(&dependency_tree) {
        println!("cargo:rerun-if-changed={}", file);
      }
    }

    Ok(Self {
      dependency_tree,
      options,
      content_hash,
    })
  }

  fn iter_files_to_watch(dep_tree: &DependencyTree) -> impl Iterator<Item = String> {
    dep_tree
      .all_files_including_dependencies()
      .into_iter()
      .map(|path| path.to_string())
  }

  fn get_contents_hash(options: &WgslBindgenOption, dep_tree: &DependencyTree) -> String {
    let mut hasher = blake3::Hasher::new();

    hasher.update(format!("{:?}", options).as_bytes());
    hasher.update(PKG_VER.as_bytes());

    for SourceFile { content, .. } in dep_tree.parsed_files() {
      hasher.update(content.as_bytes());
    }

    hasher.finalize().to_string()
  }

  fn generate_naga_module_for_entry(
    entry: SourceWithFullDependenciesResult<'_>,
  ) -> Result<WgslEntryResult, WgslBindgenError> {
    let map_err = |composer: &Composer, err: ComposerError| {
      let msg = err.emit_to_string(composer);
      WgslBindgenError::NagaModuleComposeError {
        entry: entry.source_file.file_path.to_string(),
        inner: err.inner,
        msg,
      }
    };

    let mut composer = Composer::default();
    let source = entry.source_file;

    for dependency in entry.full_dependencies.iter() {
      composer
        .add_composable_module(ComposableModuleDescriptor {
          source: &dependency.content,
          file_path: &dependency.file_path.to_string(),
          language: ShaderLanguage::Wgsl,
          as_name: dependency.module_name.as_ref().map(|name| name.to_string()),
          ..Default::default()
        })
        .map(|_| ())
        .map_err(|err| map_err(&composer, err))?;
    }

    let module = composer
      .make_naga_module(NagaModuleDescriptor {
        source: &source.content,
        file_path: &source.file_path.to_string(),
        ..Default::default()
      })
      .map_err(|err| map_err(&composer, err))?;

    Ok(WgslEntryResult {
      mod_name: source.file_path.file_prefix(),
      naga_module: module,
      source_including_deps: entry,
    })
  }

  pub fn generate_string(&self) -> Result<String, WgslBindgenError> {
    use std::fmt::Write;
    let entry_results = self
      .dependency_tree
      .get_source_files_with_full_dependencies()
      .into_iter()
      .map(Self::generate_naga_module_for_entry)
      .collect::<Result<Vec<_>, _>>()?;

    let mut text = String::new();

    if !self.options.skip_header_comments {
      writeln!(&mut text, "// File automatically generated by {PKG_NAME}^").unwrap();
      writeln!(&mut text, "//").unwrap();
      writeln!(&mut text, "// ^ {PKG_NAME} version {PKG_VER}",).unwrap();
      writeln!(&mut text, "// Changes made to this file will not be saved.").unwrap();
      writeln!(&mut text, "// SourceHash: {}", self.content_hash).unwrap();
      writeln!(&mut text).unwrap();
    }

    let output = create_rust_bindings(entry_results, &self.options)?;
    text += &output;

    Ok(text)
  }

  pub fn generate(&self) -> Result<(), WgslBindgenError> {
    let output_path = self
      .options
      .output_file
      .as_ref()
      .ok_or(WgslBindgenError::OutputFileNotSpecified)?;

    let old_content =
      std::fs::read_to_string(output_path).unwrap_or_else(|_| String::new());

    let old_hashstr_comment = old_content
      .lines()
      .find(|line| line.starts_with("// SourceHash:"))
      .unwrap_or("");

    let is_hash_changed =
      || old_hashstr_comment != format!("// SourceHash: {}", &self.content_hash);

    if self.options.skip_hash_check || is_hash_changed() {
      let content = self.generate_string()?;
      std::fs::File::create(output_path)?.write_all(content.as_bytes())?
    }

    Ok(())
  }
}
