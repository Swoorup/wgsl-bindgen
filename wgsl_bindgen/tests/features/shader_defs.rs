use std::fs::read_to_string;

use miette::{IntoDiagnostic, Result};
use syn::parse_str;
use wgsl_bindgen::{assert_tokens_snapshot, *};

#[test]
fn test_shader_defs_basic() -> Result<()> {
  let shader_defs = vec![
    ("USE_TIME".to_string(), ShaderDefValue::Bool(true)),
    ("USE_SCALE".to_string(), ShaderDefValue::Bool(true)),
    ("DEBUG_MODE".to_string(), ShaderDefValue::Bool(false)),
  ];

  WgslBindgenOptionBuilder::default()
    .workspace_root("tests/shaders/features/shader_defs")
    .entry_points(vec!["tests/shaders/features/shader_defs/test_shader.wgsl".to_string()])
    .skip_hash_check(true)
    .serialization_strategy(WgslTypeSerializeStrategy::Bytemuck)
    .type_map(GlamWgslTypeMap)
    .shader_source_type(WgslShaderSourceType::EmbedSource)
    .add_shader_defs(shader_defs)
    .derive_serde(false)
    .emit_rerun_if_change(false)
    .skip_header_comments(true)
    .output("tests/output/features/shader_defs_basic.actual.rs")
    .build()?
    .generate()
    .into_diagnostic()?;

  let actual =
    read_to_string("tests/output/features/shader_defs_basic.actual.rs").unwrap();
  let parsed_output = parse_str(&actual).unwrap();
  assert_tokens_snapshot!(parsed_output);
  assert_rust_compilation!(parsed_output);
  Ok(())
}

#[test]
fn test_shader_defs_with_texture() -> Result<()> {
  let shader_defs = vec![
    ("USE_TIME".to_string(), ShaderDefValue::Bool(true)),
    ("USE_SCALE".to_string(), ShaderDefValue::Bool(true)),
    ("USE_TEXTURE".to_string(), ShaderDefValue::Bool(true)),
    ("DEBUG_MODE".to_string(), ShaderDefValue::Bool(false)),
  ];

  WgslBindgenOptionBuilder::default()
    .workspace_root("tests/shaders/features/shader_defs")
    .entry_points(vec!["tests/shaders/features/shader_defs/test_shader.wgsl".to_string()])
    .skip_hash_check(true)
    .serialization_strategy(WgslTypeSerializeStrategy::Bytemuck)
    .type_map(GlamWgslTypeMap)
    .shader_source_type(WgslShaderSourceType::ComposerWithRelativePath)
    .add_shader_defs(shader_defs)
    .derive_serde(false)
    .emit_rerun_if_change(false)
    .skip_header_comments(true)
    .output("tests/output/features/shader_defs_with_texture.actual.rs")
    .build()?
    .generate()
    .into_diagnostic()?;

  let actual =
    read_to_string("tests/output/features/shader_defs_with_texture.actual.rs").unwrap();
  let parsed_output = parse_str(&actual).unwrap();
  assert_tokens_snapshot!(parsed_output);
  assert_rust_compilation!(parsed_output);
  Ok(())
}

#[test]
fn test_shader_defs_minimal() -> Result<()> {
  // Test with minimal shader_defs (no optional features enabled)
  let shader_defs = vec![("DEBUG_MODE".to_string(), ShaderDefValue::Bool(false))];

  WgslBindgenOptionBuilder::default()
    .workspace_root("tests/shaders/features/shader_defs")
    .entry_points(vec!["tests/shaders/features/shader_defs/test_shader.wgsl".to_string()])
    .skip_hash_check(true)
    .serialization_strategy(WgslTypeSerializeStrategy::Bytemuck)
    .type_map(GlamWgslTypeMap)
    .shader_source_type(WgslShaderSourceType::ComposerWithRelativePath)
    .add_shader_defs(shader_defs)
    .derive_serde(false)
    .emit_rerun_if_change(false)
    .skip_header_comments(true)
    .output("tests/output/features/shader_defs_minimal.actual.rs")
    .build()?
    .generate()
    .into_diagnostic()?;

  let actual =
    read_to_string("tests/output/features/shader_defs_minimal.actual.rs").unwrap();
  let parsed_output = parse_str(&actual).unwrap();
  assert_tokens_snapshot!(parsed_output);
  assert_rust_compilation!(parsed_output);
  Ok(())
}

#[test]
fn test_shader_defs_builder_methods() -> Result<()> {
  // Test the builder helper methods
  WgslBindgenOptionBuilder::default()
    .workspace_root("tests/shaders/features/shader_defs")
    .entry_points(vec!["tests/shaders/features/shader_defs/test_shader.wgsl".to_string()])
    .skip_hash_check(true)
    .serialization_strategy(WgslTypeSerializeStrategy::Bytemuck)
    .type_map(GlamWgslTypeMap)
    .shader_source_type(WgslShaderSourceType::ComposerWithRelativePath)
    .add_shader_def("USE_TIME", ShaderDefValue::Bool(true))
    .add_shader_def("USE_SCALE", ShaderDefValue::Bool(false))
    .add_shader_def("DEBUG_MODE", ShaderDefValue::Bool(true))
    .derive_serde(false)
    .emit_rerun_if_change(false)
    .skip_header_comments(true)
    .output("tests/output/features/shader_defs_builder_methods.actual.rs")
    .build()?
    .generate()
    .into_diagnostic()?;

  let actual =
    read_to_string("tests/output/features/shader_defs_builder_methods.actual.rs")
      .unwrap();
  let parsed_output = parse_str(&actual).unwrap();
  assert_tokens_snapshot!(parsed_output);
  assert_rust_compilation!(parsed_output);
  Ok(())
}
