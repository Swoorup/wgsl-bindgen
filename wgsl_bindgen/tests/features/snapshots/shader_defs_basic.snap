---
source: wgsl_bindgen/tests/features/shader_defs.rs
---
#![allow(unused, non_snake_case, non_camel_case_types, non_upper_case_globals)]
#[derive(Clone, Copy, Debug, PartialEq, Eq, Hash)]
pub enum ShaderEntry {
  TestShader,
}
impl ShaderEntry {
  pub fn create_pipeline_layout(&self, device: &wgpu::Device) -> wgpu::PipelineLayout {
    match self {
      Self::TestShader => test_shader::create_pipeline_layout(device),
    }
  }
  pub fn create_shader_module_embed_source(
    &self,
    device: &wgpu::Device,
  ) -> wgpu::ShaderModule {
    match self {
      Self::TestShader => test_shader::create_shader_module_embed_source(device),
    }
  }
}
mod _root {
  pub use super::*;
  pub trait SetBindGroup {
    fn set_bind_group(
      &mut self,
      index: u32,
      bind_group: &wgpu::BindGroup,
      offsets: &[wgpu::DynamicOffset],
    );
  }
  impl SetBindGroup for wgpu::RenderPass<'_> {
    fn set_bind_group(
      &mut self,
      index: u32,
      bind_group: &wgpu::BindGroup,
      offsets: &[wgpu::DynamicOffset],
    ) {
      self.set_bind_group(index, bind_group, offsets);
    }
  }
  impl SetBindGroup for wgpu::RenderBundleEncoder<'_> {
    fn set_bind_group(
      &mut self,
      index: u32,
      bind_group: &wgpu::BindGroup,
      offsets: &[wgpu::DynamicOffset],
    ) {
      self.set_bind_group(index, bind_group, offsets);
    }
  }
}
pub mod layout_asserts {
  use super::{_root, _root::*};
  const WGSL_BASE_TYPE_ASSERTS: () = {
    assert!(std::mem::size_of::<glam::IVec2>() == 8);
    assert!(std::mem::align_of::<glam::IVec2>() == 4);
    assert!(std::mem::size_of::<glam::IVec3>() == 12);
    assert!(std::mem::align_of::<glam::IVec3>() == 4);
    assert!(std::mem::size_of::<glam::IVec4>() == 16);
    assert!(std::mem::align_of::<glam::IVec4>() == 4);
    assert!(std::mem::size_of::<glam::UVec2>() == 8);
    assert!(std::mem::align_of::<glam::UVec2>() == 4);
    assert!(std::mem::size_of::<glam::UVec3>() == 12);
    assert!(std::mem::align_of::<glam::UVec3>() == 4);
    assert!(std::mem::size_of::<glam::UVec4>() == 16);
    assert!(std::mem::align_of::<glam::UVec4>() == 4);
    assert!(std::mem::size_of::<glam::Vec2>() == 8);
    assert!(std::mem::align_of::<glam::Vec2>() == 4);
    assert!(std::mem::size_of::<glam::Vec3>() == 12);
    assert!(std::mem::align_of::<glam::Vec3>() == 4);
    assert!(std::mem::size_of::<glam::Vec4>() == 16);
    assert!(std::mem::align_of::<glam::Vec4>() == 16);
    assert!(std::mem::size_of::<glam::Mat2>() == 16);
    assert!(std::mem::align_of::<glam::Mat2>() == 16);
    assert!(std::mem::size_of::<glam::Mat3A>() == 48);
    assert!(std::mem::align_of::<glam::Mat3A>() == 16);
    assert!(std::mem::size_of::<glam::Mat4>() == 64);
    assert!(std::mem::align_of::<glam::Mat4>() == 16);
  };
  const TEST_SHADER_UNIFORMS_ASSERTS: () = {
    assert!(std::mem::offset_of!(test_shader::Uniforms, color) == 0);
    assert!(std::mem::offset_of!(test_shader::Uniforms, time) == 16);
    assert!(std::mem::offset_of!(test_shader::Uniforms, scale) == 20);
    assert!(std::mem::size_of::<test_shader::Uniforms>() == 32);
  };
}
pub mod test_shader {
  use super::{_root, _root::*};
  #[repr(C, align(16))]
  #[derive(Debug, PartialEq, Clone, Copy)]
  pub struct Uniforms {
    #[doc = "offset: 0, size: 16, type: `vec4<f32>`"]
    pub color: glam::Vec4,
    #[doc = "offset: 16, size: 4, type: `f32`"]
    pub time: f32,
    #[doc = "offset: 20, size: 4, type: `f32`"]
    pub scale: f32,
    pub _pad_scale: [u8; 0x8],
  }
  impl Uniforms {
    pub const fn new(color: glam::Vec4, time: f32, scale: f32) -> Self {
      Self {
        color,
        time,
        scale,
        _pad_scale: [0; 0x8],
      }
    }
  }
  #[repr(C)]
  #[derive(Debug, PartialEq, Clone, Copy)]
  pub struct UniformsInit {
    pub color: glam::Vec4,
    pub time: f32,
    pub scale: f32,
  }
  impl UniformsInit {
    pub fn build(&self) -> Uniforms {
      Uniforms {
        color: self.color,
        time: self.time,
        scale: self.scale,
        _pad_scale: [0; 0x8],
      }
    }
  }
  impl From<UniformsInit> for Uniforms {
    fn from(data: UniformsInit) -> Self {
      data.build()
    }
  }
  pub const ENTRY_VS_MAIN: &str = "vs_main";
  pub const ENTRY_FS_MAIN: &str = "fs_main";
  #[derive(Debug)]
  pub struct VertexEntry<const N: usize> {
    pub entry_point: &'static str,
    pub buffers: [wgpu::VertexBufferLayout<'static>; N],
    pub constants: Vec<(&'static str, f64)>,
  }
  pub fn vertex_state<'a, const N: usize>(
    module: &'a wgpu::ShaderModule,
    entry: &'a VertexEntry<N>,
  ) -> wgpu::VertexState<'a> {
    wgpu::VertexState {
      module,
      entry_point: Some(entry.entry_point),
      buffers: &entry.buffers,
      compilation_options: wgpu::PipelineCompilationOptions {
        constants: &entry.constants,
        ..Default::default()
      },
    }
  }
  pub fn vs_main_entry() -> VertexEntry<0> {
    VertexEntry {
      entry_point: ENTRY_VS_MAIN,
      buffers: [],
      constants: Default::default(),
    }
  }
  #[derive(Debug)]
  pub struct FragmentEntry<const N: usize> {
    pub entry_point: &'static str,
    pub targets: [Option<wgpu::ColorTargetState>; N],
    pub constants: Vec<(&'static str, f64)>,
  }
  pub fn fragment_state<'a, const N: usize>(
    module: &'a wgpu::ShaderModule,
    entry: &'a FragmentEntry<N>,
  ) -> wgpu::FragmentState<'a> {
    wgpu::FragmentState {
      module,
      entry_point: Some(entry.entry_point),
      targets: &entry.targets,
      compilation_options: wgpu::PipelineCompilationOptions {
        constants: &entry.constants,
        ..Default::default()
      },
    }
  }
  pub fn fs_main_entry(targets: [Option<wgpu::ColorTargetState>; 1]) -> FragmentEntry<1> {
    FragmentEntry {
      entry_point: ENTRY_FS_MAIN,
      targets,
      constants: Default::default(),
    }
  }
  #[derive(Debug)]
  pub struct WgpuBindGroup0EntriesParams<'a> {
    pub uniforms: wgpu::BufferBinding<'a>,
  }
  #[derive(Clone, Debug)]
  pub struct WgpuBindGroup0Entries<'a> {
    pub uniforms: wgpu::BindGroupEntry<'a>,
  }
  impl<'a> WgpuBindGroup0Entries<'a> {
    pub fn new(params: WgpuBindGroup0EntriesParams<'a>) -> Self {
      Self {
        uniforms: wgpu::BindGroupEntry {
          binding: 0,
          resource: wgpu::BindingResource::Buffer(params.uniforms),
        },
      }
    }
    pub fn into_array(self) -> [wgpu::BindGroupEntry<'a>; 1] {
      [self.uniforms]
    }
    pub fn collect<B: FromIterator<wgpu::BindGroupEntry<'a>>>(self) -> B {
      self.into_array().into_iter().collect()
    }
  }
  #[derive(Debug)]
  pub struct WgpuBindGroup0(wgpu::BindGroup);
  impl WgpuBindGroup0 {
    pub const LAYOUT_DESCRIPTOR: wgpu::BindGroupLayoutDescriptor<'static> =
      wgpu::BindGroupLayoutDescriptor {
        label: Some("TestShader::BindGroup0::LayoutDescriptor"),
        entries: &[
          #[doc = " @binding(0): \"uniforms\""]
          wgpu::BindGroupLayoutEntry {
            binding: 0,
            visibility: wgpu::ShaderStages::VERTEX_FRAGMENT,
            ty: wgpu::BindingType::Buffer {
              ty: wgpu::BufferBindingType::Uniform,
              has_dynamic_offset: false,
              min_binding_size: std::num::NonZeroU64::new(std::mem::size_of::<
                _root::test_shader::Uniforms,
              >() as _),
            },
            count: None,
          },
        ],
      };
    pub fn get_bind_group_layout(device: &wgpu::Device) -> wgpu::BindGroupLayout {
      device.create_bind_group_layout(&Self::LAYOUT_DESCRIPTOR)
    }
    pub fn from_bindings(device: &wgpu::Device, bindings: WgpuBindGroup0Entries) -> Self {
      let bind_group_layout = Self::get_bind_group_layout(device);
      let entries = bindings.into_array();
      let bind_group = device.create_bind_group(&wgpu::BindGroupDescriptor {
        label: Some("TestShader::BindGroup0"),
        layout: &bind_group_layout,
        entries: &entries,
      });
      Self(bind_group)
    }
    pub fn set(&self, pass: &mut impl SetBindGroup) {
      pass.set_bind_group(0, &self.0, &[]);
    }
  }
  #[doc = " Bind groups can be set individually using their set(render_pass) method, or all at once using `WgpuBindGroups::set`."]
  #[doc = " For optimal performance with many draw calls, it's recommended to organize bindings into bind groups based on update frequency:"]
  #[doc = "   - Bind group 0: Least frequent updates (e.g. per frame resources)"]
  #[doc = "   - Bind group 1: More frequent updates"]
  #[doc = "   - Bind group 2: More frequent updates"]
  #[doc = "   - Bind group 3: Most frequent updates (e.g. per draw resources)"]
  #[derive(Debug, Copy, Clone)]
  pub struct WgpuBindGroups<'a> {
    pub bind_group0: &'a WgpuBindGroup0,
  }
  impl<'a> WgpuBindGroups<'a> {
    pub fn set(&self, pass: &mut impl SetBindGroup) {
      self.bind_group0.set(pass);
    }
  }
  #[derive(Debug)]
  pub struct WgpuPipelineLayout;
  impl WgpuPipelineLayout {
    pub fn bind_group_layout_entries(
      entries: [wgpu::BindGroupLayout; 1],
    ) -> [wgpu::BindGroupLayout; 1] {
      entries
    }
  }
  pub fn create_pipeline_layout(device: &wgpu::Device) -> wgpu::PipelineLayout {
    device.create_pipeline_layout(&wgpu::PipelineLayoutDescriptor {
      label: Some("TestShader::PipelineLayout"),
      bind_group_layouts: &[&WgpuBindGroup0::get_bind_group_layout(device)],
      push_constant_ranges: &[],
    })
  }
  pub fn create_shader_module_embed_source(device: &wgpu::Device) -> wgpu::ShaderModule {
    let source = std::borrow::Cow::Borrowed(SHADER_STRING);
    device.create_shader_module(wgpu::ShaderModuleDescriptor {
      label: Some("test_shader.wgsl"),
      source: wgpu::ShaderSource::Wgsl(source),
    })
  }
  pub const SHADER_STRING: &str = r#"
struct Uniforms {
    color: vec4<f32>,
    time: f32,
    scale: f32,
}

@group(0) @binding(0) 
var<uniform> uniforms: Uniforms;

@vertex 
fn vs_main() -> @builtin(position) vec4<f32> {
    var pos: vec4<f32> = vec4<f32>(0f, 0f, 0f, 1f);

    let _e4 = pos;
    let _e7 = uniforms.scale;
    pos = (_e4 * _e7);
    let _e11 = pos.x;
    let _e14 = uniforms.time;
    pos.x = (_e11 + sin(_e14));
    let _e17 = pos;
    return _e17;
}

@fragment 
fn fs_main() -> @location(0) vec4<f32> {
    var color: vec4<f32>;

    let _e2 = uniforms.color;
    color = _e2;
    color = vec4<f32>(1f, 0f, 0f, 1f);
    let _e9 = color;
    return _e9;
}
"#;
}
pub mod bytemuck_impls {
  use super::{_root, _root::*};
  unsafe impl bytemuck::Zeroable for test_shader::Uniforms {}
  unsafe impl bytemuck::Pod for test_shader::Uniforms {}
}
